helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,167 @@
+      Map<String, ? extends HelixProperty> newPropertyMap) {
+    Collection<String> changedItems = new HashSet<>();
+    oldPropertyMap.forEach((name, property) -> {
+      if (newPropertyMap.containsKey(name)
+          && !property.getRecord().equals(newPropertyMap.get(name).getRecord())) {
+        changedItems.add(name);
+      }
+    });
+    return changedItems;
+  }
+
+  /**
+   * Return a collection of names that are newly added.
+   * @return
+   */
+  private Collection<String> getAddedItems(Map<String, ? extends HelixProperty> oldPropertyMap,
+      Map<String, ? extends HelixProperty> newPropertyMap) {
+    return Sets.difference(newPropertyMap.keySet(), oldPropertyMap.keySet());
+  }
+
+  /**
+   * Return a collection of names that were removed.
+   * @return
+   */
+  private Collection<String> getRemovedItems(Map<String, ? extends HelixProperty> oldPropertyMap,
+      Map<String, ? extends HelixProperty> newPropertyMap) {
+    return Sets.difference(oldPropertyMap.keySet(), newPropertyMap.keySet());
+  }
+
+  private void clearCachedComputation() {
+    _changedItems.clear();
+    _addedItems.clear();
+    _removedItems.clear();
+  }
+
+  /**
+   * Initializes old and new snapshots when ResourceChangeDetector gets its first update.
+   */
+  private void initializeSnapshots() {
+    _oldSnapshot = new ResourceChangeSnapshot();
+    _newSnapshot = new ResourceChangeSnapshot();
+  }
+
+  /**
+   * Based on the change type given and propertyMap type, call the right getters for propertyMap.
+   * @param changeType
+   * @param snapshot
+   * @return
+   */
+  private Map<String, ? extends HelixProperty> determinePropertyMapByType(
+      HelixConstants.ChangeType changeType, ResourceChangeSnapshot snapshot) {
+    switch (changeType) {
+    case INSTANCE_CONFIG:
+      return snapshot.getInstanceConfigMap();
+    case IDEAL_STATE:
+      return snapshot.getIdealStateMap();
+    case RESOURCE_CONFIG:
+      return snapshot.getResourceConfigMap();
+    case LIVE_INSTANCE:
+      return snapshot.getLiveInstances();
+    default:
+      throw new HelixException(String.format(
+          "ResourceChangeDetector cannot compute the names of changes for the given ChangeType: %s",
+          changeType));
+    }
+  }
+
+  /**
+   * Makes the current newSnapshot the oldSnapshot and reads in the up-to-date snapshot for change
+   * computation. To be called in the controller pipeline.
+   * @param dataProvider newly refreshed DataProvider (cache)
+   */
+  public synchronized void updateSnapshots(ResourceControllerDataProvider dataProvider) {
+    // If snapshots are null, initialize them
+    if (_oldSnapshot == null || _newSnapshot == null) {
+      initializeSnapshots();
+    }
+
+    // If there are changes, update internal states
+    _oldSnapshot = new ResourceChangeSnapshot(_newSnapshot);
+    _newSnapshot = new ResourceChangeSnapshot(dataProvider);
+    dataProvider.clearRefreshedChangeTypes();
```

Comment:
This method has been called in the ResourceChangeSnapshot.
In addition, although I don't have a strong preference, how about having a getAndClearRefreshedChangeTypes()?

Reply:
We agreed to cache the results between `updateSnapshots()` calls. If you did `getAndClearRefreshedChangeTypes()`, then getChangeTypes() won't be idempotent between update calls! Do you want to introduce that inconsistency?

Reply:
I don't. The thing is that you call it in the ResourceChangeSnapshot constructor. So if you create 2 snapshots using the same data provider, the 2nd one will miss the change, right?

My point is that, either you remove the clearRefreshedChangeTypes() call in ResourceChangeSnapshot constructor. Or we redesign the logic with getAndClearRefreshedChangeTypes.
The first option would be simpler.

Reply:
I get your comment now @jiajunwang. Thanks for clarifying. Yes, there was a duplicate call. I removed it in the constructor :)